barretenberg/cpp/src/barretenberg/stdlib/primitives/safe_uint/safe_uint.test.cpp

```diff
@@ -186,30 +212,76 @@ TYPED_TEST(SafeUintTest, TestSubtractMethodUnderflowFails)
         c = c.subtract(d, suint_ct::MAX_BIT_NUM);
         FAIL() << "Expected out of range error";
     } catch (std::runtime_error const& err) {
+        EXPECT_TRUE(composer.check_circuit());
         EXPECT_EQ(err.what(), std::string("maximum value exceeded in safe_uint subtract"));
     } catch (...) {
         FAIL() << "Expected std::runtime_error modulus in safe_uint class";
     }
 }
 #endif
+
+/**
+ * @brief Test that underflow is caught in general case.
+ * @details General case refers to when difference.current_max + other.current_max does not exceed MAX_VALUE.
+ */
+#if !defined(__wasm__)
+TYPED_TEST(SafeUintTest, TestSubtractMethodUnderflowFailsGeneral)
+{
+    STDLIB_TYPE_ALIASES
+    auto composer = Composer();
+
+    field_ct a(witness_ct(&composer, 0));
+    field_ct b(witness_ct(&composer, 1));
+    suint_ct c(a, 0);
+    suint_ct d(b, 1);
+    c = c.subtract(d, suint_ct::MAX_BIT_NUM);
+    EXPECT_FALSE(composer.check_circuit());
+}
+#endif
+
 // - OPERATOR
 
-TYPED_TEST(SafeUintTest, TestMinusOperator)
+/**
+ * @brief Test that valid minus operation works.
+ */
+TYPED_TEST(SafeUintTest, TestMinusOperator0)
+{
+    STDLIB_TYPE_ALIASES
+    auto composer = Composer();
+
+    field_ct a(witness_ct(&composer, 9));
+    field_ct b(witness_ct(&composer, 2));
+    suint_ct c(a, 4);
+    suint_ct d(b, 2);
+    c = c - d; // 9 - 2 = 7 should not underflow
+
+    EXPECT_TRUE(composer.check_circuit());
+}
+
+/**
+ * @brief Test that valid minus operation works on 0.
+ */
+#if !defined(__wasm__)
+TYPED_TEST(SafeUintTest, TestMinusOperator1)
 {
     STDLIB_TYPE_ALIASES
     auto composer = Composer();
 
     field_ct a(witness_ct(&composer, 2));
-    field_ct b(witness_ct(&composer, 9));
+    field_ct b(witness_ct(&composer, 2));
     suint_ct c(a, 2);
-    suint_ct d(b, 4);
-    c = d - c;
-
+    suint_ct d(b, 3);
+    c = c - d; // 2 - 2 = 0 should not overflow, even if d has more bits than c.
     EXPECT_TRUE(composer.check_circuit());
 }
+#endif
 
+/**
+ * @brief Test that checks that minus operator underflow is caught in the general case.
+ * @details General case means that the special case does not happen.
+ */
 #if !defined(__wasm__)
-TYPED_TEST(SafeUintTest, TestMinusOperatorUnderflowFails)
+TYPED_TEST(SafeUintTest, TestMinusOperatorUnderflowFailsGeneral0)
```

Comment:
interesting proposal, I agree that the current names are not very good, but also am wary about merging multiple tests into one test, because I feel like tests should be split to be as small as possible.

Reply:
we have several places in the codebase where we do several test iterations (e.g. failing and successful) in one test, i found 2 bugs since i'm here thanks to them.  I don't have strong opinions about how you structure tests but it'd be really helpful to give them some other names and contain their length as much as possible (but i know naming is hard..)

Reply:
Renamed, let me know if its still not satisfactory